### PR TITLE
feat: pushdown table and partition key predicates to catalog (#736)

### PR DESCRIPTION
**Pull request by tustvold**
Relates to #736 and https://github.com/influxdata/influxdb_iox/issues/735 #734

This plumbs predicate pushdown into the catalog, and provides implementations for pruning based on partition key and table name.  Pruning based on chunk data is left for a future PR. The major benefit of pushing predicates through to the catalog is it allows pruning to occur before snapshotting any chunks.

It should be noted that the way the SQL frontend is currently wired up means that no predicates are passed down, this is tracked by #734

### CODE REVIEW

File: query/src/lib.rs [hunk 1]
```diff
@@ -59,16 +58,6 @@ pub trait PartitionChunk: Debug + Send + Sync {
     /// particular partition.
     fn id(&self) -> u32;
 
-    /// Returns true if this chunk *might* have data that passes the
-    /// predicate. If false is returned, this chunk can be
-    /// skipped entirely. If true is returned, there still may not be
-    /// rows that match.
-    ///
-    /// This is used during query planning to skip including entire chunks
-    fn could_pass_predicate(&self, _predicate: &Predicate) -> Result<bool, Self::Error> {
```
Comment (tustvold) [inline query/src/lib.rs:68]: Predicate evaluation now takes place prior to snapshotting to a DbChunk (which is what implements PartitionChunk). This method can therefore be removed

File: server/src/db/catalog.rs [hunk 2]
```diff
@@ -189,6 +185,33 @@ impl Catalog {
...
+            Some(partition_key) => {
+                itertools::Either::Right(partitions.get(partition_key).into_iter())
+            }
+        };
+
+        for partition in partitions {
+            let partition = partition.read();
+            chunks.extend(partition.filtered_chunks(predicate).map(|chunk| {
+                let chunk = chunk.read();
+                // TODO: Filter chunks
```
Comment (tustvold) [inline server/src/db/catalog.rs:209]: This is where per-chunk predicates would be evaluated if supported

File: server/src/db/chunk.rs [hunk 3]
```diff
@@ -290,30 +290,6 @@ impl PartitionChunk for DbChunk {
         }
     }
 
-    fn could_pass_predicate(&self, _predicate: &Predicate) -> Result<bool> {
```
Comment (tustvold) [inline server/src/db/chunk.rs:293]: This will instead be implemented on the catalog chunks

File: query/src/frontend/influxrpc.rs [hunk 4]
```diff
@@ -1175,43 +1175,11 @@ impl InfluxRpcPlanner {
         &self,
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:1175]: It seems like we could remove the `filtered_chunks` function entirely now (as the filtering is done inside the Database::chunks call )

File: query/src/lib.rs [hunk 5]
```diff
@@ -42,10 +42,9 @@ pub trait Database: Debug + Send + Sync {
     /// Return the partition keys for data in this DB
     fn partition_keys(&self) -> Result<Vec<String>, Self::Error>;
 
-    /// Returns a covering set of chunks in the specified partition. A
-    /// covering set means that together the chunks make up a single
-    /// complete copy of the data being queried.
-    fn chunks(&self, partition_key: &str) -> Vec<Arc<Self::Chunk>>;
+    /// Returns a set of chunks within the partition that match
+    /// the provided predicate
```
Comment (alamb) [inline query/src/lib.rs:46]:
```suggestion
    /// Returns a set of chunks within the partition with data that may match
    /// the provided predicate. If possible, chunks which have no rows that can 
    /// possibly match the predicate are omitted. 
```

File: server/src/db/catalog/partition.rs [hunk 6]
```diff
@@ -180,6 +181,23 @@ impl Partition {
         self.tables.values().flat_map(|table| table.chunks.values())
     }
 
+    /// Return an iterator over chunks in this partition that
+    /// may pass the provided predicate
+    pub fn filtered_chunks<'a>(
+        &'a self,
+        predicate: &'a Predicate,
+    ) -> impl Iterator<Item = &Arc<RwLock<Chunk>>> + 'a {
+        match &predicate.table_names {
```
Comment (alamb) [inline server/src/db/catalog/partition.rs:190]: I wonder if using `Predicate::should_include_table` might be cleaner here (and also allow for easier future expansion of "does a table pass a predicate")?

Something like (untested)
```
 self.tables()
  .iter()
  .filter_map((|table_name, table|  {
    if predicate.should_include_table(table_name) { 
      Some(table.chunks())
    } else {
      None
    }
  })
  .flatten()

```

**Review by alamb — APPROVED** [review]
I think this is a nice change @tustvold 

I would suggest calling this PR something more specific like "skip chunks without the relevant tables" because, as you note there are many other scenarios we might want to prune on in the future. 

However, I think the structure here is really nice and it is a good step forward 👍

**Pull request merged**